File: day-10-collisions-load-factor/python/hash-table.py
def simple_hash(key, size):
    total = sum(ord(ch) for ch in key)
    return total % size

class SimpleHashTable:
    def __init__(self, size=10):
        self.size = size
        self.table = [[] for _ in range(size)] # List of buckets

    def put(self, key, value):
        idx = simple_hash(key, self.size)
        bucket = self.table[idx]
        # update if key exists
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket[i] = (key,  value)
                return
        bucket.append((key, value))

    def get(self, key):
        idx = simple_hash(key, self.size)
        bucket = self.table[idx]
        for k, v in bucket:
            if k == key:
                return v
        return None
    
    def display(self):
        for i, bucket in enumerate(self.table):
            print(f"Slot {i}: {bucket}")

ht = SimpleHashTable(5)
ht.put("cat", "meow")
ht.put("dog", "woof")
ht.put("act", "meow again")
print(ht.get("cat"))
print(ht.get("act"))
ht.display()

File: day-10-collisions-load-factor/js/hash-table.js
// Simple hash function: sum char codes modulo size
function hash(key, size) {
  let total = 0;
  for (let i = 0; i < key.length; i++) {
    total += key.charCodeAt(i);
  }
  return total % size;
}

class simpleHashTable {
  constructor(size = 10) {
    this.table = new Array(size);
    for (let i = 0; i < size; i++) {
      this.table[i] = []; // Each slot is an array (bucket)
    }
  }

  put(key, value) {
    const index = hash(key, this.table.length);
    const bucket = this.table[index];
    for (let pair of bucket) {
      if (pair[0] === key) {
        pair[1] = value;
        return;
      }
    }
    // Not found, push new pair
    bucket.push([key, value]);
  }

  get(key) {
    const index = hash(key, this.table.length);
    const bucket = this.table[index];
    for (let pair of bucket) {
      if (pair[0] === key) {
        return pair[1];
      }
    }
    return undefined;
  }
  //   Helper to see the interval state
  display() {
    for (let i = 0; i < this.table.length; i++) {
      console.log(`Slot ${i}: ${JSON.stringify(this.table[i])}`);
    }
  }
}

const ht = new simpleHashTable(5);
ht.put("cat", "meow");
ht.put("dog", "woof");
ht.put("act", "meow again!");
console.log(ht.get("cat"));
console.log(ht.get("act"));

ht.display();
